### PR TITLE
[Enhancement] Allow one-phase local aggregate in the single BE

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Problem Summary：

If there is only one BE and one fragment instance, we could generate one-phase local aggregation with the local shuffle operator (Scan->OnePhaseAgg->LocalShuffle) regardless of the differences between grouping keys and scan distribution keys.

## Modification
- Add `withLocalShuffle` to `LogicalAggregationOperator` and `PhysicalAggregationOperator`.
- Add `GenerateAggregateWithLocalShuffleRule` to set `LogicalAggregationOperator#withLocalShuffle` as true.
- `HashDistributionSpec#isSatisfy()` returns true when required property need local shuffle, and this is `LOCAL` scan.
- `PlanFragmentBuilder#visitPhysicalHashAggregate()` set `isNeedLocalShuffle` and `AssignScanRangesPerDriverSeq` according to `PhysicalAggregationOperator#withLocalShuffle`.

### CODE REVIEW

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.List;
+
+/**
+ * If there is only one BE and one fragment instance, we could generate one-phase local aggregation
+ * with the local shuffle operator (Scan->OnePhaseAgg->LocalShuffle) regardless of the differences
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java:19]: Scan->LocalShuffle->OnePhaseAgg

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                Pattern.create(OperatorType.LOGICAL_AGGR, OperatorType.PATTERN_LEAF));
+    }
+
+    private static final GenerateAggregateWithLocalShuffleRule instance = new GenerateAggregateWithLocalShuffleRule();
+
+    public static GenerateAggregateWithLocalShuffleRule getInstance() {
+        return instance;
+    }
+
+    public static boolean enable(SessionVariable sessionVariable) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java:35]: Need we disable the two phase aggregate in the memo?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 3]
```diff
@@ -145,6 +145,8 @@ public class SessionVariable implements Serializable, Writable, Cloneable {
     // them do the real work on core.
     public static final String ENABLE_PIPELINE = "enable_pipeline";
 
+    public static final String ENABLE_LOCAL_SHUFFLE_AGG = "enable_local_shuffle_agg";
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:148]: add some comments

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/HashDistributionDesc.java [hunk 4]
```diff
@@ -30,12 +31,21 @@ public enum SourceType {
     // Which operator produce this hash DistributionDesc
     private final SourceType sourceType;
 
+    private final boolean withLocalShuffle;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/HashDistributionDesc.java:34]: Add some comments for local shuffle. most fo Optimizer guys don't know it.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java (hunk 2 above)
Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java:35]: Maybe not, when the aggregation cardinality is very low, two-phase is better than one-phase with local shuffle.

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 80 / 81 (98.77%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/HashDistributionDesc.java|8|9|88.89%|[131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java|15|15|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/HashAggImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/PropertyDeriverBase.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/HashDistributionSpec.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|5|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java [hunk 5]
```diff
@@ -261,6 +263,14 @@ public void setAssignScanRangesPerDriverSeq(boolean assignScanRangesPerDriverSeq
         this.assignScanRangesPerDriverSeq = assignScanRangesPerDriverSeq;
     }
 
+    public boolean isNeedLocalShuffleOperator() {
+        return needLocalShuffleOperator;
+    }
+
+    public void setNeedLocalShuffleOperator(boolean needLocalShuffleOperator) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/planner/PlanFragment.java:270]: Could we show local shuffle info in explain if the local shuffle is true?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 86 / 87 (98.85%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/HashDistributionDesc.java|8|9|88.89%|[131]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/GenerateAggregateWithLocalShuffleRule.java|15|15|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/HashAggImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/AggregationNode.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashAggregateOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalAggregationOperator.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/PropertyDeriverBase.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/HashDistributionSpec.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|5|5|100.00%|[]|

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
The enhancement point is handle the sence about high cardinality aggregate on one BE node, and will compare cost of single-stage aggregate plan and multi-stage aggregate plan, but we conside some question:
* when the optimize return a single-stage aggregate with exchange,  means it's must a high cardinality aggregate, not low cardinality aggregate, so to compare the cost of local-shuffle mode is meaningless
* the enhancement only work when scan on one BE node, but the optimizer don't care the actually execute node now, so optimizer don't know the scan only work on one BE node, I think the optimizer need to consider this in the future and need a complete solution, but the way of this pr is too intrusive for optimizer framework.

So I have a suggestion, you can modify on FragmentBuilder:
* can be modify then the Fragment generate way
* add mark on exchange node to do local-shuffle
* mark the BE node on ScanNode, and check isn't only one BE node when meet aggregate node.

and consider the point 1, you don't care the optimizer will return a multi-stage aggregate plan when low cardinality aggregate.

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 6]
```diff
@@ -1249,9 +1249,20 @@ public PlanFragment visitPhysicalHashAggregate(OptExpression optExpr, ExecPlan c
...
+
+            if (node.isOnePhaseAgg() && hasNoExchangeNodes(inputFragment.getPlanRoot())) {
+                boolean withLocalShuffle = node.isWithLocalShuffleOperator();
+                if (withLocalShuffle) {
+                    inputFragment.setEnableSharedScan(true);
+                    inputFragment.setAssignScanRangesPerDriverSeq(false);
+                } else {
+                    inputFragment.setEnableSharedScan(false);
+                    inputFragment.setAssignScanRangesPerDriverSeq(true);
+                }
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1263]: inputFragment.setEnableSharedScan(withLocalShuffle);
inputFragment.setAssignScanRangesPerDriverSeq(!withLocalShuffle);

**Pull request closed without merging**